### DISCUSSION TITLE
Compress selected images with Tinify

### DISCUSSION THREAD

**Discussion by vojtaholik** [johnlindquist/kit#328] [Share]
![tinify](https://user-images.githubusercontent.com/25487857/123306748-f706ff00-d521-11eb-9f11-6744d7b1e76b.gif)

```js
// Menu: Tinify
// Description: Compress selected images with Tinify
// Author: Vojta Holik
// Twitter: @vjthlk

let tinify = await npm("tinify");
let fs = await import("fs");
let selectedFiles = await getSelectedFile();

tinify.key = await env("TINIFY_API_KEY", {
  hint: md("get your Tinify api key at https://tinypng.com/developers"),
  ignoreBlur: true,
  secret: true,
});

let filePaths = selectedFiles.split("\n");

for (let filePath of filePaths) {
  let directory = path.dirname(filePath);
  let extension = path.extname(filePath);
  let originalFileName = path.basename(filePath);
  let suffix = "";
  let isHD = originalFileName.includes("@2x");
  let newFileName = isHD
    ? originalFileName
        .replace("@2x", "")
        .replace(extension, `${suffix}@2x${extension}`)
    : originalFileName.replace(extension, `${suffix}${extension}`);

  fs.readFile(filePath, (err, sourceData) => {
    if (err) throw err;
    tinify.fromBuffer(sourceData).toBuffer((err, resultData) => {
      if (err) throw err;
      fs.writeFile(`${directory}/` + newFileName, resultData, (err) => {
        if (err) throw err;
      });
    });
  });
}

```

---

<details>
  <summary>You can change file names and directory to best suite your workflow. Since I often work with <code>@2x</code> images I adjusted the script to correctly suffix that part.</summary>

```js
let isHD = originalFileName.includes("@2x");
let newFileName = isHD
  ? originalFileName
      .replace("@2x", "")
      .replace(extension, `${suffix}@2x${extension}`)
  : originalFileName.replace(extension, `${suffix}${extension}`);
```

</details>

**Comment by johnlindquist** [comment]
Looks great, awesome work @vojtaholik !

When writing a script to share, I recommend using the `env` helper:

```js
tinify.key = await env("TINIFY_API_KEY", {
  ignoreBlur: true,
  hint: md(
    `Get your api key on [https://tinify.com/dashboard/api](tinify.com/dashboard/api)`
  ),
})
```
![CleanShot 2021-06-26 at 10 00 47](https://user-images.githubusercontent.com/36073/123518916-66d1d280-d665-11eb-992e-8faa5a6854c5.png)

This only pops up when the var isn't set. It's a nice way of documenting where to grab env vars from. The link is even clickable 😇


## fs

Both `await readFile` and `await writeFile` are global. So you use those if you like. It's my preference to try and avoid callbacks in scripts and rely on pausing execution with `await`, but that's just a personal preference. It really makes no difference.

**Comment by dealingwith** [comment]
is `getSelectedFile` deprecated? I'm getting an error

**Reply by dealingwith** [reply]
Disregard!
